File: 297SerializeandDeserializeBinaryTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Codec {

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        if(root == null) return "#";
        StringBuilder sb = new StringBuilder();
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        while(!q.isEmpty()) {
            TreeNode node = q.poll();
            if (node == null){
                sb.append("#,");
                continue;
            }
            sb.append(node.val + ",");
            q.add(node.left);
            q.add(node.right);
        }
        return sb.toString();
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        if(data == "#") return null;
        Queue<TreeNode> q = new LinkedList<>();
        String[] vals = data.split(",");
        TreeNode root = new TreeNode(Integer.parseInt(vals[0]));
        q.add(root);
        for (int i = 1; i < vals.length; i++) {
            TreeNode node = q.poll();
            if(!vals[i].equals("#")) {
                node.left = new TreeNode(Integer.parseInt(vals[i]));
                q.add(node.left);
            }
            if(!vals[++i].equals("#")) {
                node.right = new TreeNode(Integer.parseInt(vals[i]));
                q.add(node.right);
            }   
        }
        
        return root;
    }

}

File: 297SerializeandDeserializeBinaryTree.cpp
class Codec {
public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        if (!root) return "#";
        return to_string(root->val) + "," + serialize(root->left)+ ","+ serialize(root->right);
        
        
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
       TreeNode* root = helper(data);
        return root;
        
    }
    
    TreeNode* helper(string& data) {
        if (data[0] == '#'){
            if (data.size()>1) {
                data = data.substr(2);
            }
            return NULL;
        }
        TreeNode* node = new TreeNode(getVal(data));
        node->left = helper (data);
        node->right = helper(data);
        return node;   
    }
    
    int getVal(string& data){
        int pos = data.find(',');
        int val = stoi(data.substr(0,pos));
        data = data.substr(pos+1);
        return val;    
    }
    
    
};

